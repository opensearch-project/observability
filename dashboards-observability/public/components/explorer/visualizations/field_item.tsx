/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import './field_item.scss';

import React, { useState } from 'react';
import {
  EuiIconTip,
  EuiPopover
} from '@elastic/eui';
import { i18n } from '@osd/i18n';
import { EuiHighlight } from '@elastic/eui';
import { FieldButton } from '../../common/field_button';
import { DragDrop } from './drag_drop';
import { LensFieldIcon } from './lens_field_icon';

import { debouncedComponent } from '../../common/debounced_component';

function wrapOnDot(str?: string) {
  // u200B is a non-width white-space character, which allows
  // the browser to efficiently word-wrap right after the dot
  // without us having to draw a lot of extra DOM elements, etc
  return str ? str.replace(/\./g, '.\u200B') : '';
}

export const InnerFieldItem = function InnerFieldItem(props: any) {
  const {
    field,
    highlight,
    exists,
    hideDetails,
  } = props;

  const [infoIsOpen, setOpen] = useState(false);

  function togglePopover() {}

  const lensFieldIcon = <LensFieldIcon type={field.type} />;
  const lensInfoIcon = (
    <EuiIconTip
      anchorClassName="lnsFieldItem__infoIcon"
      content={
        hideDetails
          ? i18n.translate('xpack.lens.indexPattern.fieldItemTooltip', {
              defaultMessage: 'Drag and drop to visualize.',
            })
          : exists
          ? i18n.translate('xpack.lens.indexPattern.fieldStatsButtonLabel', {
              defaultMessage: 'Click for a field preview, or drag and drop to visualize.',
            })
          : i18n.translate('xpack.lens.indexPattern.fieldStatsButtonEmptyLabel', {
              defaultMessage:
                'This field doesnâ€™t have any data but you can still drag and drop to visualize.',
            })
      }
      type="iInCircle"
      color="subdued"
      size="s"
    />
  );
  return (
    <EuiPopover
      ownFocus
      className="lnsFieldItem__popoverAnchor"
      display="block"
      data-test-subj="lnsFieldListPanelField"
      container={document.querySelector<HTMLElement>('.application') || undefined}
      button={
        <DragDrop
          label={field.displayName}
          data-test-subj={`lnsFieldListPanelField-${field.name}`}
          draggable
        >
          <FieldButton
            className={`lnsFieldItem lnsFieldItem--${field.type} lnsFieldItem--${
              exists ? 'exists' : 'missing'
            }`}
            isActive={infoIsOpen}
            onClick={togglePopover}
            aria-label={i18n.translate('xpack.lens.indexPattern.fieldStatsButtonAriaLabel', {
              defaultMessage: '{fieldName}: {fieldType}. Hit enter for a field preview.',
              values: {
                fieldName: field.displayName,
                fieldType: field.type,
              },
            })}
            fieldIcon={lensFieldIcon}
            fieldName={
              <EuiHighlight search={wrapOnDot(highlight)}>
                {wrapOnDot(field.name)}
              </EuiHighlight>
            }
            fieldInfoIcon={lensInfoIcon}
          />
        </DragDrop>
      }
      isOpen={infoIsOpen}
      closePopover={() => setOpen(false)}
      anchorPosition="rightUp"
      panelClassName="lnsFieldItem__fieldPanel"
    >
    </EuiPopover>
  );
};

export const FieldItem = debouncedComponent(InnerFieldItem);
