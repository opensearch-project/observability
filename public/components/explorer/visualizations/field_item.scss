/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

.lnsFieldItem {
  .lnsFieldItem__infoIcon {
    visibility: hidden;
    opacity: 0;
  }

  &:hover:not([class*='isActive']) {
    cursor: grab;

    .lnsFieldItem__infoIcon {
      visibility: visible;
      opacity: 1;
      transition: opacity $euiAnimSpeedFast ease-in-out 1s;
    }
  }
}

.lnsFieldItem--missing {
  background: lightOrDarkTheme(transparentize($euiColorMediumShade, .9), $euiColorEmptyShade);
  color: $euiColorDarkShade;
}

.lnsFieldItem__topValue {
  margin-bottom: $euiSizeS;

  &:last-of-type {
    margin-bottom: 0;
  }
}

.lnsFieldItem__topValueProgress {
  background-color: $euiColorLightestShade;

  // sass-lint:disable-block no-vendor-prefixes
  &::-webkit-progress-bar {
    background-color: $euiColorLightestShade;
  }
}

.lnsFieldItem__fieldPanel {
  min-width: 260px;
  max-width: 300px;
}

.lnsFieldItem__buttonGroup {
  // Enforce lowercase for buttons or else some browsers inherit all caps from flyout title
  text-transform: none;
}
