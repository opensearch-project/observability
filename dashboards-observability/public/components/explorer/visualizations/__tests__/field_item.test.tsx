/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { configure, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import React from 'react';
import { waitFor } from '@testing-library/react';
import { FieldItem } from '../field_item';
import { 
  AVAILABLE_FIELDS
} from '../../../../../test/event_analytics_constants';

describe('Visualization field item component', () => {
  configure({ adapter: new Adapter() });

  it('Renders field item component', async () => {
    
    const wrapper = mount(
      <FieldItem
        field={AVAILABLE_FIELDS[0]}
        highlight=""
        exists={true}
        hideDetails={false}
        key={AVAILABLE_FIELDS[0].name}
      />
    );
    
    wrapper.update();

    await waitFor(() => {
      expect(wrapper).toMatchSnapshot();
    });
  });
});
