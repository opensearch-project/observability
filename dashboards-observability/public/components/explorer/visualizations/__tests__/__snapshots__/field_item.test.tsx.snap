// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Visualization field item component Renders field item component 1`] = `
<Component
  exists={true}
  field={
    Object {
      "name": "agent",
      "type": "string",
    }
  }
  hideDetails={false}
  highlight=""
>
  <Memo(InnerFieldItem)
    exists={true}
    field={
      Object {
        "name": "agent",
        "type": "string",
      }
    }
    hideDetails={false}
    highlight=""
  >
    <EuiPopover
      anchorPosition="rightUp"
      button={
        <DragDrop
          data-test-subj="lnsFieldListPanelField-agent"
          draggable={true}
        >
          <FieldButton
            aria-label=": string. Hit enter for a field preview."
            className="lnsFieldItem lnsFieldItem--string lnsFieldItem--exists"
            fieldIcon={
              <LensFieldIcon
                type="string"
              />
            }
            fieldInfoIcon={
              <EuiIconTip
                anchorClassName="lnsFieldItem__infoIcon"
                color="subdued"
                content="Click for a field preview, or drag and drop to visualize."
                size="s"
                type="iInCircle"
              />
            }
            fieldName={
              <EuiHighlight
                search=""
              >
                agent
              </EuiHighlight>
            }
            isActive={false}
            onClick={[Function]}
          />
        </DragDrop>
      }
      className="lnsFieldItem__popoverAnchor"
      closePopover={[Function]}
      data-test-subj="lnsFieldListPanelField"
      display="block"
      hasArrow={true}
      isOpen={false}
      ownFocus={true}
      panelClassName="lnsFieldItem__fieldPanel"
      panelPaddingSize="m"
    >
      <EuiOutsideClickDetector
        isDisabled={true}
        onOutsideClick={[Function]}
      >
        <div
          className="euiPopover euiPopover--anchorRightUp euiPopover--displayBlock lnsFieldItem__popoverAnchor"
          data-test-subj="lnsFieldListPanelField"
          onKeyDown={[Function]}
          onMouseDown={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchStart={[Function]}
        >
          <div
            className="euiPopover__anchor"
          >
            <DragDrop
              data-test-subj="lnsFieldListPanelField-agent"
              draggable={true}
            >
              <Memo(DragDropInner)
                data-test-subj="lnsFieldListPanelField-agent"
                draggable={true}
                isDragging={true}
                isNotDroppable={false}
                setDragging={[Function]}
              >
                <FieldButton
                  aria-label=": string. Hit enter for a field preview."
                  className="lnsFieldItem lnsFieldItem--string lnsFieldItem--exists lnsDragDrop lnsDragDrop-isDraggable lnsDragDrop-isDragging"
                  data-test-subj="lnsFieldListPanelField-agent"
                  draggable={true}
                  fieldIcon={
                    <LensFieldIcon
                      type="string"
                    />
                  }
                  fieldInfoIcon={
                    <EuiIconTip
                      anchorClassName="lnsFieldItem__infoIcon"
                      color="subdued"
                      content="Click for a field preview, or drag and drop to visualize."
                      size="s"
                      type="iInCircle"
                    />
                  }
                  fieldName={
                    <EuiHighlight
                      search=""
                    >
                      agent
                    </EuiHighlight>
                  }
                  isActive={false}
                  onClick={[Function]}
                  onDragEnd={[Function]}
                  onDragLeave={[Function]}
                  onDragOver={[Function]}
                  onDragStart={[Function]}
                  onDrop={[Function]}
                >
                  <div
                    aria-label=": string. Hit enter for a field preview."
                    className="osdFieldButton lnsFieldItem lnsFieldItem--string lnsFieldItem--exists lnsDragDrop lnsDragDrop-isDraggable lnsDragDrop-isDragging"
                    data-test-subj="lnsFieldListPanelField-agent"
                    draggable={true}
                    onDragEnd={[Function]}
                    onDragLeave={[Function]}
                    onDragOver={[Function]}
                    onDragStart={[Function]}
                    onDrop={[Function]}
                  >
                    <button
                      className="osd-resetFocusState osdFieldButton__button"
                      onClick={[Function]}
                    >
                      <span
                        className="osdFieldButton__fieldIcon"
                      >
                        <LensFieldIcon
                          type="string"
                        >
                          <FieldIcon
                            className="lnsFieldListPanel__fieldIcon"
                            type="string"
                          >
                            <EuiToken
                              aria-label="string"
                              className="osdFieldIcon lnsFieldListPanel__fieldIcon"
                              iconType="tokenString"
                              size="s"
                              title="string"
                            >
                              <span
                                className="euiToken euiToken--euiColorVis1 euiToken--square euiToken--light euiToken--small osdFieldIcon lnsFieldListPanel__fieldIcon"
                                style={Object {}}
                              >
                                <EuiIcon
                                  aria-label="string"
                                  size="m"
                                  title="string"
                                  type="tokenString"
                                >
                                  <EuiIconEmpty
                                    aria-hidden={true}
                                    aria-label="string"
                                    className="euiIcon euiIcon--medium euiIcon-isLoading"
                                    focusable="false"
                                    role="img"
                                    style={null}
                                    title="string"
                                  >
                                    <svg
                                      aria-hidden={true}
                                      aria-label="string"
                                      className="euiIcon euiIcon--medium euiIcon-isLoading"
                                      focusable="false"
                                      height={16}
                                      role="img"
                                      style={null}
                                      viewBox="0 0 16 16"
                                      width={16}
                                      xmlns="http://www.w3.org/2000/svg"
                                    />
                                  </EuiIconEmpty>
                                </EuiIcon>
                              </span>
                            </EuiToken>
                          </FieldIcon>
                        </LensFieldIcon>
                      </span>
                      <span
                        className="osdFieldButton__name"
                      >
                        <EuiHighlight
                          search=""
                        >
                          <span>
                            agent
                          </span>
                        </EuiHighlight>
                      </span>
                      <div
                        className="osdFieldButton__infoIcon"
                      >
                        <EuiIconTip
                          anchorClassName="lnsFieldItem__infoIcon"
                          color="subdued"
                          content="Click for a field preview, or drag and drop to visualize."
                          size="s"
                          type="iInCircle"
                        >
                          <EuiToolTip
                            anchorClassName="lnsFieldItem__infoIcon"
                            content="Click for a field preview, or drag and drop to visualize."
                            delay="regular"
                            position="top"
                          >
                            <span
                              className="euiToolTipAnchor lnsFieldItem__infoIcon"
                              onKeyUp={[Function]}
                              onMouseOut={[Function]}
                              onMouseOver={[Function]}
                            >
                              <EuiIcon
                                aria-label="Info"
                                color="subdued"
                                onBlur={[Function]}
                                onFocus={[Function]}
                                size="s"
                                tabIndex={0}
                                type="iInCircle"
                              >
                                <EuiIconEmpty
                                  aria-hidden={true}
                                  aria-label="Info"
                                  className="euiIcon euiIcon--small euiIcon--subdued euiIcon-isLoading"
                                  focusable="true"
                                  onBlur={[Function]}
                                  onFocus={[Function]}
                                  role="img"
                                  style={null}
                                  tabIndex={0}
                                >
                                  <svg
                                    aria-hidden={true}
                                    aria-label="Info"
                                    className="euiIcon euiIcon--small euiIcon--subdued euiIcon-isLoading"
                                    focusable="true"
                                    height={16}
                                    onBlur={[Function]}
                                    onFocus={[Function]}
                                    role="img"
                                    style={null}
                                    tabIndex={0}
                                    viewBox="0 0 16 16"
                                    width={16}
                                    xmlns="http://www.w3.org/2000/svg"
                                  />
                                </EuiIconEmpty>
                              </EuiIcon>
                            </span>
                          </EuiToolTip>
                        </EuiIconTip>
                      </div>
                    </button>
                  </div>
                </FieldButton>
              </Memo(DragDropInner)>
            </DragDrop>
          </div>
        </div>
      </EuiOutsideClickDetector>
    </EuiPopover>
  </Memo(InnerFieldItem)>
</Component>
`;
